### PR TITLE
feat(AHWR-4): enable and use Pino logger

### PR DESCRIPTION
**Pull request by mattyod**
**Logging with [hapi-pino](https://www.npmjs.com/package/hapi-pino)**

All requests are now logged and parsed to JSON so that deeply nested items are preserved and not replaced with `[Object]` or `[Array]`.

**Before:**
Screenshot 2024-10-15 at 5 20 47 PM

**After:**
Screenshot 2024-10-15 at 4 33 00 PM

**Added to all log entries:**

* consistent timestamps
* log levels: info, warn, error etcetera
* app name, e.g. `ffc-ahwr-farmer-apply`
* response statusCode
* other [Pino](https://www.npmjs.com/package/pino) defaults

**setBindings():**

During a request lifecycle a logger instance is available on `request.logger`. Bindings (or defaults for the instance) can be added and will then be used whenever the logger is used. e.g.
```
request.logger.setBindings({ example: 'binding' })
request.logger.info('hello') // { example: 'binding', message: 'hello', ... }
```

**env settings:**

Logs are silenced for tests based on `process.env.NODE_ENV`.

Pretty logging can be enabled via `USE_PRETTY_PRINT`
Screenshot 2024-10-15 at 4 42 32 PM

### CODE REVIEW

File: app/lib/email/notify-client.js
Comment (aaroncarroll82) [inline app/lib/email/notify-client.js:1]: I think the answer is no...but are these deleted files definitely not needed? Seems to be some legacy notification for access being granted or not which I assume is from the days when there were eligibility checks

File: app/messaging/application/index.js [hunk 1]
```diff
@@ -8,8 +8,6 @@ async function getApplication (applicationReference, sessionId) {
 }
 
 async function sendApplication (application, sessionId) {
-  console.log(`Sending application ${JSON.stringify(application)} to queue ${applicationRequestQueue.address} with sessionID ${sessionId}.`)
```
Comment (aaroncarroll82) [inline app/messaging/application/index.js:11]: Maybe worth keeping this line but via proper logger

**Review by aaroncarroll82 — COMMENTED** [review]
My other comments would be -
1. How does this look like when viewed on Azure logs? I guess maybe we won't know till it's deployed.
2. Are the request specific bindings cleared/reset between requests? We're not going to run into a situation whereby some request sets an sbi or something, and then following logs that are not related still have this in context?

File: app/plugins/logging.js [hunk 2]
```diff
@@ -1,11 +1,45 @@
...
+  isBoom: err.isBoom,
+  isServer: err.isServer,
+  payload: err.payload,
+  stack: err.stack,
+  data: {
+    isResponseError: err?.data?.isResponseError,
+    payload: err?.data?.payload
+  }
+})
+process.env.USE_PRETTY_PRINT = 'true'
```
[the remark this answers is not in the document]
Comment (mattyod) [reply app/plugins/logging.js]: thanks good spot. Removed.

File: app/plugins/logging.js [hunk 3]
```diff
@@ -1,11 +1,45 @@
-const config = require('../config')
+const pino = require('hapi-pino')
```
[the remark this answers is not in the document]
Comment (mattyod) [reply app/plugins/logging.js:1]: agree I think, but maybe not for this PR?

File: app/api-requests/business-eligible-to-apply.js [hunk 4]
```diff
@@ -9,23 +9,19 @@ const { CannotReapplyTimeLimitError, OutstandingAgreementError, AlreadyAppliedEr
...
-    return
+    return 'no existing applications'
   }
   const latestApplication = getLatestApplication(latestApplicationsForSbi)
   if (validStatusForApplication.includes(latestApplication.statusId)) {
-    // latest application is either WITHDRAWN or NOT_AGREED so okay to continue
-    console.log(`${new Date().toISOString()} Business is eligible to apply : ${JSON.stringify({
-        sbi: latestApplication.data.organisation.sbi
-      })}`)
+    return 'latest application either WITHDRAWN or NOT AGREED'
```
[the remark this answers is not in the document]
Comment (mattyod) [reply app/api-requests/business-eligible-to-apply.js]: done

File: app/auth/auth-code-grant/redeem-authorization-code-for-access-token.js [hunk 5]
```diff
@@ -30,15 +29,10 @@ const redeemAuthorizationCodeForAccessToken = async (request) => {
         json: true
       }
     )
-    if (response.res.statusCode !== 200) {
```
[the remark this answers is not in the document]
Comment (mattyod) [reply app/auth/auth-code-grant/redeem-authorization-code-for-access-token.js:33]: I suspect all these statusCode checks are a legacy from when the app maybe used a fetch client in the past? 

[Wreck throws an error for any response status <400](https://github.com/hapijs/wreck/blob/master/lib/index.js#L574), so this code can never run. Unless we get a 201 or something, and then it probably shouldn't anyway.

File: app/auth/token-verify/acquire-signing-key.js [hunk 6]
```diff
@@ -2,23 +2,14 @@ const Wreck = require('@hapi/wreck')
...
 const acquireSigningKey = async () => {
-  console.log(`${new Date().toISOString()} Acquiring the signing key data necessary to validate the signature`)
-  try {
-    const response = await Wreck.get(
-      `${config.authConfig.defraId.hostname}/discovery/v2.0/keys?p=${config.authConfig.defraId.policy}`,
-      {
-        json: true
-      }
-    )
-    if (response.res.statusCode !== 200) {
```
[the remark this answers is not in the document]
Comment (mattyod) [reply app/auth/token-verify/acquire-signing-key.js:13]: Do we want to catch responses between other than 200 but less than 400? If it's 400 or above, Wreck throws and we catch higher up the chain. If we error accessing `response.payload.keys[0]` a similar story but we'll get a stack trace.

```
{
  "level": "error",
  "time": 1729070265347,
  "pid": 323,
  "hostname": "cc3eb8de8dcb",
  "name": "ffc-ahwr-farmer-apply",
  "req": {
    "id": "1729070264472:cc3eb8de8dcb:323:m2bns4mz:10001",
    "method": "get",
    "url": "/apply/signin-oidc"
  },
  "tempApplicationId": "TEMP-D1F9-4866",
  "err": {
    "name": "TypeError",
    "message": "Cannot read properties of undefined (reading 'banana')",
    "stack": "TypeError: Cannot read properties of undefined (reading 'banana')
      at acquireSigningKey (/home/node/app/auth/token-verify/acquire-signing-key.js:12:26)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async jwtVerify (/home/node/app/auth/token-verify/jwt-verify.js:6:15)
      at async Object.authenticate (/home/node/app/auth/authenticate.js:18:3)
      at async handler (/home/node/app/routes/signin-oidc.js:60:9)
      at async exports.Manager.execute (/home/node/node_modules/@hapi/hapi/lib/toolkit.js:60:28)
      at async internals.handler (/home/node/node_modules/@hapi/hapi/lib/handler.js:46:20)
      at async exports.execute (/home/node/node_modules/@hapi/hapi/lib/handler.js:31:20)
      at async Request._lifecycle (/home/node/node_modules/@hapi/hapi/lib/request.js:371:32)
      at async Request._execute (/home/node/node_modules/@hapi/hapi/lib/request.js:281:9)",
    "data": {}
  },
  "msg": "check details"
}
```

File: app/lib/email/notify-client.js
Comment (mattyod) [reply app/lib/email/notify-client.js:1]: I'll ask someone else to confirm @alex742, @govTechSatish  but I couldn't find anything that was referencing them.

File: app/messaging/application/index.js (hunk 1 above)
Comment (mattyod) [reply app/messaging/application/index.js:11]: Don't think we want to be logging the sessionId or the users application (as it contains personal details). What I've done here is put the temporary application id onto the logger before we make this call, and the new one afterwards if successful. https://github.com/DEFRA/ffc-ahwr-farmer-apply/pull/424/files#diff-678633002b18fd55567e0abbbfbbb8f6dd06d19f3824dfb27e88425de276d29dR50-R83

So if this errors you'll get something like:
```
{
  "level": "error",
  "time": 1729080472899,
  "pid": 95,
  "hostname": "02708ebe0a24",
  "name": "ffc-ahwr-farmer-apply",
  "req": {
    "id": "1729080471237:02708ebe0a24:95:m2bttint:10028",
    "method": "post",
    "url": "/apply/declaration"
  },
  "tempApplicationReference": "TEMP-4251-9621",
  "sbi": "107167406",
  "statusCode": 500,
  "message": "An internal server error occurred",
  "stack": "Error: Apply declaration returned a null application reference.
    at handler (/home/node/app/routes/declaration.js:89:20)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async exports.Manager.execute (/home/node/node_modules/@hapi/hapi/lib/toolkit.js:60:28)
    at async internals.handler (/home/node/node_modules/@hapi/hapi/lib/handler.js:46:20)
    at async exports.execute (/home/node/node_modules/@hapi/hapi/lib/handler.js:31:20)
    at async Request._lifecycle (/home/node/node_modules/@hapi/hapi/lib/request.js:371:32)
    at async Request._execute (/home/node/node_modules/@hapi/hapi/lib/request.js:281:9)",
  "msg": "pre response error"
}
```

And if it succeeds you'll see something like:

```
{
  "level": "info",
  "time": 1729083940385,
  "pid": 303,
  "hostname": "02708ebe0a24",
  "name": "ffc-ahwr-farmer-apply",
  "req": {
    "id": "1729083938284:02708ebe0a24:303:m2bvvts4:10061",
    "method": "post",
    "url": "/apply/endemics/declaration"
  },
  "tempApplicationReference": "TEMP-72B2-1616",
  "sbi": "107167406",
  "newApplicationReference": "IAHW-63B1-8D68",
  "res": { "statusCode": 200 },
  "responseTime": 2101,
  "msg": "[response] post /apply/endemics/declaration 200 (2101ms)"
}
```

File: app/api-requests/rpa-api/base.js [hunk 7]
```diff
@@ -8,23 +8,16 @@ const get = async (hostname, url, request, headers = {}) => {
   const token = session.getToken(request, tokens.accessToken)
   headers[apiHeaders.xForwardedAuthorization] = token
   headers[apiHeaders.ocpSubscriptionKey] = config.authConfig.apim.ocpSubscriptionKey
-  console.log(`${new Date().toISOString()} Request message to RPA: ` + JSON.stringify(`${hostname}${url}`))
```
[the remark this answers is not in the document]
Comment (mattyod) [reply app/api-requests/rpa-api/base.js:11]: This 'base' file is used by a bunch of others so I felt the fact that we were logging 'doing what I always do, with properties that do not change' wasn't useful. All the RPA API calls are part of the check user authentication step for this handler https://github.com/DEFRA/ffc-ahwr-farmer-apply/pull/424/files#diff-187dd3107f0ebf909c35530e40789124043d380bc1a8eb3456931d282103d339R57 so what I've done is update the logger bindings as we progress, temp-id, authenticated: true, actual-id, info about the previous application.

```
{
  "level": "info",
  "time": 1729085451352,
  "pid": 318,
  "hostname": "02708ebe0a24",
  "name": "ffc-ahwr-farmer-apply",
  "req": {
    "id": "1729085448954:02708ebe0a24:318:m2bw1xtv:10041",
    "method": "get",
    "url": "/apply/signin-oidc"
  },
  "tempApplicationId": "TEMP-D5C5-166C",
  "iss": "https://dcidmtest.b2clogin.com/131a35fb-0422-49c9-8753-15217cec5411/v2.0/",
  "authenticated": true,
  "sbi": 107167406,
  "previousApplication": "no existing applications",
  "userType": "farmerApply",
  "msg": "logged in user"
}
```

[the remark this answers is not in the document]
Comment (aaroncarroll82) [reply app/api-requests/rpa-api/base.js:11]: That's OK, with the request logging it's obvious the requests are being made so no need for the extra line

File: app/auth/token-verify/acquire-signing-key.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (aaroncarroll82) [reply app/auth/token-verify/acquire-signing-key.js:13]: OK I think we can live with that for now.

File: app/lib/email/notify-client.js
Comment (aaroncarroll82) [reply app/lib/email/notify-client.js:1]: Will mark as resolved if we get feedback they aren;t used. If that is the case, might also be a good idea to good boy scout it, and remove those values from the config too whilst there

File: app/messaging/application/index.js (hunk 1 above)
Comment (aaroncarroll82) [reply app/messaging/application/index.js:11]: Definitely would not want the pii bits logged out, but didn;t know if a log line saying sent to queue provided value. To be fair thinking about it, it probably doesn;t add anything

File: app/lib/email/notify-client.js
Comment (mattyod) [reply app/lib/email/notify-client.js:1]: aah good point, will do (assuming nothing else uses them).

Comment (mattyod) [reply app/lib/email/notify-client.js:1]: @govTechSatish has confirmed that these are no longer used, cleaned up the remaining notify config and references.

**Review by aaroncarroll82 — APPROVED** [review]

**Pull request merged**